### ISSUE TITLE
5.0: Narrowing from type-guards ignored when constructing JSX call (2604)

### ISSUE THREAD

**Issue by eps1lon** [microsoft/TypeScript#53178]
# Bug Report

### 🔎 Search Terms

does not have any construct or call signatures.(2604)

### 🕗 Version & Regression Information

- This changed between versions 4.9.5 and 5.0.1-rc (I'm pretty sure it used to work in 5.0.0-beta but can't verify with playground)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.1.0-dev.20230309#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0D09cAYgDbIwAywqMACsTCoA2mW68AEjgAmWKAGESkAHZZlMMgF0as3O0pxcEZbzgBZZAE8ARlnEwpyWQqXG1MAFyYqMAHSLwN3UAFUswLAAeZGVLAD44AB9vPD9sFIBRViwQd0S4ZQBXVlY8guVZAmBVaR0sPWQDIxN4e0dnAJV3LzT8f1dVELDI6Li8ntSfTOz3WvrG41MAOQbiAHcsaVaZOQ6gz2Te3YGYUPComNjZ-SxDBfgpnPUAeSgJCF4jroOJjKzH+CSvCgVQA5nRGHAAFIAZQAGnAsH9cjAhuQLDY7DwHNsXIFjmQ4NIIFh0MoIPAABbIABuNxGt2aUAK+Dg0EMyGKcFQwBByg4BUoqF8AAoAEwANgADAAWACUNCapkR03UAEE4ABeb6+HgANQ5wGkD3cwvRti2Th2-XcsrgAH5zFZzVi2la8bkvGbMZIcZ91HAAGQBuARL0W9rW-30S4QmHw5X-OAo8LkcNuzrqAlEkn5clwKm0lAxBlA5nwNm4Dklbm8-mCkUSmXyxXwBPuABCmu1eoNRqR6mFadxGZgtodQ798C8EQnkfg0YYTDjCP78GTNzELt9c6zxNJeYLdOLLaZLIrVa5PL5MAFJIbUrlCruK5VMHkXYAhOMdah9axDcaA6zu66hjiGwEjnA0ZwF4EF7E+zQvv8AAiXYRMsUBrBscHHFBMZLnCSHIqiZCATALxvB8O6EnuuaUjSR6WCWMCnuWUDspyNbXreQpig+zbPm26jpF2349v+favsKZEUe8b5zmBMmvHJk4wSGSmUfJIHzrEQA)

### 💻 Code

We saw 8 errors in our codebase pop up. This might be a fairly common issue in React Native codebases or any codebase having types like `React.ComponentType<any> | React.ReactElement`.
```ts
import * as React from 'react'

// `@types/react-native` FlatListProps['ListHeaderComponent']
declare const ListHeaderComponent: React.ComponentType<any> | React.ReactElement

// JSX element type 'ListHeaderComponent' does not have any construct or call signatures.(2604)
const elementB = React.isValidElement(ListHeaderComponent) ? ListHeaderComponent : <ListHeaderComponent />
```

Note that this works either with either
```diff
-declare const ListHeaderComponent: React.ComponentType<any> | React.ReactElement
+declare const ListHeaderComponent: React.ComponentType<any> | React.ReactElement<unknown>
```
which requires a change in `@types/react-native`

or 

```diff
-declare function isValidElement(maybeElement: unknown): maybeElement is React.ReactElement<unknown>
+declare function isValidElement(maybeElement: unknown): maybeElement is React.ReactElement<any>
```
which requires a change in `@types/react`.

The change in `@types/react-native` makes more sense but means other type libraries need to update as well.

3rd option is making `React.ReactElement` default to `React.ReactElement<unkown>` instead of `React.ReactElement<any>` but that might be even more disruptive.

I'll fiddle with each alternative but I would appreciate if you could check if this new behavior in TypeScript is so valueable to warrant this breakage in the ecosystem.

### 🙁 Actual behavior

JSX element type 'ListHeaderComponent' does not have any construct or call signatures.(2604)

### 🙂 Expected behavior

No type error since `isValidElement` narrows to `React.ReactElement` i.e. the else branch should just be `React.ComponentType<any>`

**Comment by eps1lon** [comment]
The linked issue repros in 4.9.4 while mine does not. Could you clarify why you think it is a duplicate?

**Comment by RyanCavanaugh** [comment]
I believe the root cause here - `isValidElement` and `ReactElement` having different default type arguments, thus resulting in subtyping assymmetry, is still the same. If you have a repro that doesn't depend on React we can bisect it and verify.

**Comment by eps1lon** [comment]
So if it turns out to root cause is a dupe, you won't consider this a 5.0 regression even though it only started failing when going from 4.9 to 5.0?

**Comment by eps1lon** [comment]
[no React 4.9.5 no type errors](https://www.typescriptlang.org/play?jsx=4&ts=4.9.5#code/C4TwDgpgBAwg9gWzHAdhFwAq5oF4oAUAlFLgHxQoCuANjQFCiRQBKEAhgMbACiNECdMAA8ABQBOcMAGdSUdihAV8AbzCSZALigSpsgL716AEwica7cdABmVFNwCWqKA+kA1djQfG+AoQQh+QQxtOwBrFDgAdxQibUC-DBdZNi5eIKFhcMiYshMzCysoW3tgJxQoTit2YAhfYOACJghteCRUIWxmAB8oaWBxBxQAczjWDm56oQBuIwB6OagAMQtgABlXYF0ZAG0Acg3+gAkOU3E25DQMPYBdfPNLaE5UfqgAWXYQACMIQ+AT9hnC4dEKwRCXTo4KC9VKTDJJXrUOjQqB2UzWIYQYz3QpPF7AKB-AFA8Eg4CtUlXLBQmETdKJYA4x6VfFQABylkkUSxRNOEHOlKEFPaVK6ECZRWeKFeUwwAHlxEc4P1gVTtLD6Q0Uf1BiN5osAFIAZQAGlAElrmlA9h9vr9NsT+aqhHsoMY4BBZJECQALdgAN2gChALOlAyo3CgcHElU8ND6DmGKBqVCs0gAdAQAEwANgADAAWIj0KWvC1CACCclcHi8PnhjVtP15gKdgowJAA-O9Ps2HXyBSKhFBtE37ccB86kgAyaeVaq1WWN3vj-6T9vAYsLKDGs3lpJWg791uDiHXN0er1wX0BoOKUM6iME6Ox5HSRPJ4Cpz2Z3OF4ulgS+7AAAQtW7ieN4S4EC2JJDh2UDdrBbbwQS2hVBwi4NjBx5wWem70Nuu7mg2UCHshp5kq67qepQ15QH6gbyPegHiE+UYxpwcYJkmKZpr++ZFiWrLATAcgAIQ1pB9YMjhE4nlOm6IfOmF1NhFGKSQ2gaRuwlhiRDIACJyBhNRqbJHLiFyPK4Sh+FboapoGZaUJ7EuCpKiqG7UZedE3kxwYPuGkYvlxb4fnxP7ZoJAEiQ2PDgbWUHYe5irKsAmnKalnkZRuI4qWZ0HZelmlAA)

[no react 5.0.0-rc type errors](https://www.typescriptlang.org/play?jsx=4&ts=5.1.0-dev.20230310#code/C4TwDgpgBAwg9gWzHAdhFwAq5oF4oAUAlFLgHxQoCuANjQFCiRQBKEAhgMbACiNECdMAA8ABQBOcMAGdSUdihAV8AbzCSZALigSpsgL716AEwica7cdABmVFNwCWqKA+kA1djQfG+AoQQh+QQxtOwBrFDgAdxQibUC-DBdZNi5eIKFhcMiYshMzCysoW3tgJxQoTit2YAhfYOACJghteCRUIWxmAB8oaWBxBxQAczjWDm56oQBuIwB6OagAMQtgABlXYF0ZAG0Acg3+gAkOU3E25DQMPYBdfPNLaE5UfqgAWXYQACMIQ+AT9hnC4dEKwRCXTo4KC9VKTDJJXrUOjQqB2UzWIYQYz3QpPF7AKB-AFA8Eg4CtUlXLBQmETdKJYA4x6VfFQABylkkUSxRNOEHOlKEFPaVK6ECZRWeKFeUwwAHlxEc4P1gVTtLD6Q0Uf1BiN5osAFIAZQAGlAElrmlA9h9vr9NsT+aqhHsoMY4BBZJECQALdgAN2gChALOlAyo3CgcHElU8ND6DmGKBqVCs0gAdAQAEwANgADAAWIj0KWvC1CACCclcHi8PnhjVtP15gKdgowJAA-O9Ps2HXyBSKhFBtE37ccB86kgAyaeVaq1WWN3vj-6T9vAYsLKDGs3lpJWg791uDiHXN0er1wX0BoOKUM6iME6Ox5HSRPJ4Cpz2Z3OF4ulgS+7AAAQtW7ieN4S4EC2JJDh2UDdrBbbwQS2hVBwi4NjBx5wWem70Nuu7mg2UCHshp5kq67qepQ15QH6gbyPegHiE+UYxpwcYJkmKZpr++ZFiWrLATAcgAIQ1pB9YMjhE4nlOm6IfOmF1NhFGKSQ2gaRuwlhiRDIACJyBhNRqbJHLiFyPK4Sh+FboapoGZaUJ7EuCpKiqG7UZedE3kxwYPuGkYvlxb4fnxP7ZoJAEiQ2PDgbWUHYe5irKsAmnKalnkZRuI4qWZ0HZelmlAA)

**Comment by Andarist** [comment]
Bisected to https://github.com/microsoft/TypeScript/pull/52282

**Comment by RyanCavanaugh** [comment]
Right, this is basically the converse report of #50916

**Comment by eps1lon** [comment]
I'm not following the labelling here. It's considered a duplicate of an issue that was closed by the PR that caused it? Sounds more like you're saying this is intended behavior?

**Comment by greypants** [comment]
I'm also having a hard time trying to follow the thread here... I still have this issue on 5.0.3. Are we waiting on a change in @types/react then? Or is this supposed to be fixed already?

**Comment by noahtallen** [comment]
This issue unfortunately isn't fixed! I'm still getting errors in Typescript 5.1.6. (I'm _pretty_ sure this is detailing the same issue as the post.)

Here is a [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvAUASAEYEMBOAXFAM7D4CWAdgOYDc6AvuugPStQDKwArtlHABmUAEoRcAY2ABRADYQAthCrBC6asAj5Bk6GMkz5SlQB4ACsii4qIAHxosYfHDAliZhszYcYPKlIo4KlIKGipcXnxoCRcKCAATKEFnBVFxKQAaKGweYCgKBTAjZWAIwOCFXHjoHjA1QT8AoPySADVcWQp4uUUS81sACjhsACsIKWJURigAHygqHllZWag-asFqBIBKYmGxqRa0gx7jYH6HTCjI4IBCG4FR8eAAfgA6AAtcEgB5AHcqMzOSD4UADKAAcicLhI4KgW08LAa-mA5Ra8DgQ0eE1WVAA1lQ4P8dg99nkKCRYAgLlcePhbvc9k83p8fv9AS4tKCIXh8LD4UxEY0Uc14nBOHAlMB3tQaANqGBcsR0St9FITiUthcKMI5W0Ol11So5VQFcAtpqMJhLhBrvkTbkGJhmJh2FAACrgaAkd5wRaJbDQcL4Zy-BJZHJkkhUcHAG5YGl0u2m148hGC5Go0XiyXS2hyEgQY2mpVUuaqwy9FQWrDaqC69FF3Lmi7W23y3IpgiO52uj2QKAASU48wIIYSUAAtBPB1ApdRcdlxrgeAW0VTRRAozGoJ8AG7QXBQGjKLQUCSveM22nBdvAV5Q1yeIA) showing the error. The problem is that `object is ReactElement<Props>` does _not_ work to narrow the types, but `object is ReactElement` _does_ work.

**Comment by Andarist** [comment]
Just a note: it works when u "forward" the `P` [TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvAUASAEYEMBOAXFAM7D4CWAdgOYDc6AvuugPStQDKwArtlHABmUAEoRcAY2ABRADYQAthCrBC6asAj5Bk6GMkz5SlQB4ACsii4qIAHxosYfHDAliZhs3QATCBNkE0II8VFIUcFRQFCQAariyFN5yisrA5rYAFHDYAFZ+qmiMUAA+UFQ8srIlUCG+gtQQ3gCUxNl5UlEkouJSycZpZrYswaHA4ZHecJxwSsAAFtQ06RnUYDwF8Iil+r1GqelNDpgUwiux8Yl9qStUa8BNhxiYmPgQvPiRq+sMmMyY7FAACrgaAkOZwCreKDYaBUAjOADujQANFYqFCKMBOlhXu9Prd1gA6PD4TxAA):
```ts
type Foo = {
	bar: string;
}

// Stub of ReactElement:
interface ReactElement<P = any> {
	props: P;
}

declare function isValidElement<P>(object: {} | null | undefined): object is ReactElement<P>

function doSomething<P>(input: Foo | ReactElement<P>) {
	if (isValidElement(input)) {
		return input;
	}
	// Type should be narrowed, and it is
	return input.bar;
}
```

**Comment by rejetto** [comment]
an easy fix for the problem is to not use just ReactElement but ReactElement<unknown>  (if you don't know a more specific type)
